glue_astronomy/translators/trace.py

```diff
@@ -0,0 +1,42 @@
+from glue.config import data_translator
+from glue.core import Data, Subset
+
+from specreduce.tracing import Trace
+
+
+@data_translator(Trace)
+class TraceHandler:
+
+    def to_data(self, obj):
+        """
+        Convert a specreduce Trace object to a glue Data object
+
+        Parameters
+        ----------
+        obj : `specreduce.tracing.Trace`
+            The Trace object to convert
+        """
+        data = Data(x=obj.image[0], y=obj.trace)
+        if hasattr(obj, 'meta'):
+            data.meta.update(obj.meta)
+        data.meta['Trace'] = obj
+        return data
+
+    def to_object(self, data, attribute=None):
+        """
+        Convert a glue Data object to a Trace object.
+
+        Parameters
+        ----------
+        data : `glue.core.data.Data`
+            The data to convert to a Trace object
+        attribute : `glue.core.component_id.ComponentID`
+            The attribute to use for the Trace data
+        """
+
+        if isinstance(data, Subset):
+            raise NotImplementedError("cannot convert subset to Trace object")
+        if not isinstance(data.meta.get('Trace'), Trace):
+            raise TypeError("data is not of a valid specreduce Trace object")
+
+        return data.meta['Trace']
```

Comment:
This worries me, because it is completely ignoring what's actually in the `data` object itself. I think that's counter-intuitive because a likely eventual workflow is to *change* the `y` trace by hand, which is probably most naturally done in glue.

So I would suggest the following:
1. If there's some way that glue data objects can flag "have been modified since created", only do the next step if that one is true.
2. Do something like:
```suggestion
        data.meta['Trace'].trace = data.y
        return data.meta['Trace']
```

Now this *does* introduce an assymetry between x and y.  But for a trace I think that makes sense.

Regardless, I think either in `to_object` or the top-level class, the docstring should be explcit about whether it gets data from glue to create the object.

Reply:
I like this idea.  We can even include a check to make sure that `x` has not changed and that `y` is the correct length.  In cases where `y` has changed, we'll probably want to return an `ArrayTrace` with the modified `trace.y`.  For example: if the input trace object is a `FlatTrace`, but then the underlying data is modified once in glue, it no longer (necessarily) makes sense to return another `FlatTrace` object, right?

In theory, I don't see why `Trace` _needs_ the input image, it's just convenient to have attached, so if we can make that more flexible on the specreduce end, then we might be able to rebuild the object from scratch instead of relying on these references.  But I think in cases where the data isn't modified, it is nice to roundtrip and return the same type of trace, which at least involves storing some more information than the arrays themselves.

Reply:
Update: I now check to see if the "y"-component matches the trace, and if not, create a new `ArrayTrace` object with the same image.  It turns out that glue does not allow changing the shape of data components, so there is no need for any checks on the lengths, but I do raise an error if the "x"-component no longer matches the dispersion axis of the image.

Reply:
> I also agree that the memory impact is not serious given that, from my read of the internals of `Trace` and its sub-classes, both `obj.trace` and `obj.image` are references not copies, and this is just carrying along those references. As long as `glue` itself does not do any copying later on down the line this is probably "free".

Any data component initialising or being added to a `data` object is processed by `Component.autotyped()`, which will typically involve `np.asarray(data)` and potentially other re-casting operations. So for most `ndarray` inputs there will probably be no actual copying, but even if, this only involves `obj.trace`. The `meta` reference is just an `OrderedDict` item.

> 1. If there's some way that glue data objects can flag "have been modified since created", only do the next step if that one is true.

Wondering if the callback mechanism might be used for this, although I do not see an example where this is applied in a `Data` component so far, apart from the use in `_update_world_components`.